### PR TITLE
Magnetic lyapunov exponents

### LINKED ISSUE #52: add lyapunov exponent for magnetic particle
By reading the paper, we know that in actuality only a method that propagates the offset vectors in magnetic field needs to be added, all the reflection methods are the same!

### PR DESCRIPTION
**Pull request by lhupe**
Restructured & adapted Lyapunov exponent code to work for magnetic billiards

closes #52

### CODE REVIEW

File: src/lyapunov_spectrum.jl [hunk 1]
```diff
@@ -73,62 +77,90 @@ function specular!(p::Particle{T}, o::Union{InfiniteWall{T},FiniteWall{T}}, offs
...
+        temp = [Γ[1] + t*Γ[3], Γ[2] + t*Γ[4], Γ[3], Γ[4]]
         offset[:,k] = temp
     end
 end
 
+#magnetic
+function propagate_offset!(offset::MArray{Tuple{4,4},T}, t::T,
+                           p::MagneticParticle{T}) where T
+    ω = p.omega
+    sω = sin(ω*t)
```
Comment (Datseris) [inline src/lyapunov_spectrum.jl]: add comment to use sincos in 0.7

File: src/lyapunov_spectrum.jl [hunk 2]
```diff
@@ -73,62 +77,90 @@ function specular!(p::Particle{T}, o::Union{InfiniteWall{T},FiniteWall{T}}, offs
...
+         0.0 0.0     sω          cω]
+    for k in 1:4
+        Γ = offset[:,k]
+        offset[:,k] = J*Γ
+    end
+end
+
 """
-    relocate(p::Particle, o::Obstacle, t, offset::MArray) -> newt
+    propagate!(p::AbstractParticle{T}, t::T, offset::MArray{Tuple{4,4},T})
```
Comment (Datseris) [inline src/lyapunov_spectrum.jl]: I think we should remove the documentation strings, as they are added to the documentation strings of the `propagate!` function.

This brings unecessary information to someone reading the docstrings for their own implementations. 

Can you simply comment the docstrings? Do not remove them, but make them comments instead.

File: src/lyapunov_spectrum.jl [hunk 3]
```diff
@@ -73,62 +77,90 @@ function specular!(p::Particle{T}, o::Union{InfiniteWall{T},FiniteWall{T}}, offs
...
-    end
+    propagate_offset!(offset, tmin, p)
     return tmin
 end
 
 
+
 """
-    lyapunovspectrum!(p::Particle{T}, bt::Vector{Obstacle{T}}, t)
+    lyapunovspectrum!(p::AbstractParticle{T}, bt::Vector{Obstacle{T}}, t)
```
Comment (Datseris) [inline src/lyapunov_spectrum.jl]: This docstring is wrong... `Vector{Obstacle}` ?

File: test/lyapunov.jl [hunk 4]
```diff
@@ -51,3 +51,34 @@ if printinfo
 end
 return
 end
+
```
Comment (Datseris) [inline test/lyapunov.jl]: This test is not enough.

Please add 2 more tests: that check if λ for very small magnetic fields is the same as λ without magnetic fields. Do this for 2 different billiards of your choice, doesn't matter.

(there has to be some small average over 10 particles or so)

File: src/lyapunov_spectrum.jl (hunk 2 above)
Comment (lhupe) [reply src/lyapunov_spectrum.jl]: Should I do the same for the `resolvecollision` and `specular!` docstrings?

Comment (Datseris) [reply src/lyapunov_spectrum.jl]: Yes please, sorry I forgot.

File: test/lyapunov.jl (hunk 4 above)
Comment (Datseris) [reply test/lyapunov.jl]: Talking about tests, can you add one more test that confirms what the function does when given a pinned particle?

```
bt = billiard_sinai(;setting="periodic")
p = MagneticParticle(0.1, 0.5, -π/2, 2.0)
Λ = lyapunovspectrum!(p, bt, 10000.0)
@test Λ[1] == 0
```

**Comment by Datseris** [comment]
@lhupe there is another thing:
```
bt = billiard_sinai(setting = "periodic")
p = MagneticParticle(0.2, 0.5, -π/2, 3.0)
lyapunovspectrum(p, bt, 100.0)

Float64[4]
0.00
0.00
0.00102…
0.00102…
```
yet
```
next_collision(p, bt)

(Inf, 1)
```
How is this possible ? What does the function do?

**Pull request merged**